app/views/support/ABTest.scala

```diff
@@ -8,25 +8,54 @@ import java.nio.charset.StandardCharsets
 import com.gu.i18n.CountryGroup
 import com.gu.i18n.CountryGroup._
 import play.api.libs.json._
+import play.api.libs.functional.syntax._
 import play.api.mvc.{Cookie, Request}
 import play.twirl.api.Html
-import scala.util.Random
 
+import scala.util.Random
 import scalaz.NonEmptyList
 import views.html.fragments.giraffe.{contributeAmountButtons, contributeMessage}
-
-trait TestTrait {
-  type VariantFn
-  case class Variant(variantName: String, variantSlug: String, weight: Double, render: VariantFn, countryGroupFilter: Set[CountryGroup] = Set.empty) {
-    def testName = name
-    def testSlug = slug
-    def matches(countryGroup:CountryGroup):Boolean = countryGroupFilter.isEmpty || countryGroupFilter.contains(countryGroup)
+import views.support.AmountHighlightTest.AmountVariantData
+import views.support.CountryGroupImplicits._
+import views.support.MessageCopyTest.CopyVariantData
+
+sealed trait VariantData
+
+object VariantData {
+  implicit val formatter: Writes[VariantData] = new Writes[VariantData] {
+    def writes(o: VariantData): JsValue = {
+      o match {
+        case amount: AmountVariantData => AmountVariantData.format.writes(amount)
+        case copy: CopyVariantData => CopyVariantData.format.writes(copy)
+      }
+    }
   }
+}
 
+case class Variant(testName: String, testSlug: String, variantName: String, variantSlug: String, weight: Double, data: VariantData, countryGroupFilter: Set[CountryGroup] = Set.empty) {
+  def matches(countryGroup: CountryGroup): Boolean = countryGroupFilter.isEmpty || countryGroupFilter.contains(countryGroup)
+}
+
+object Variant {
+  implicit val writes: Writes[Variant] = (
```

Comment:
I just realised you can use the macro to do this one:

`implicit val writes = Json.writes[Variant]`

And that should work